kong/kong.lua

```diff
@@ -157,6 +161,34 @@ function Kong.ssl_certificate()
   end
 end
 
+function Kong.balancer()
+  local addr = ngx.ctx.balancer_address
+  addr.tries = addr.tries + 1
+  if addr.tries > 1 then  
+    -- only call balancer on retry, first one is done in `core.access.before` which runs
+    -- in the ACCESS context and hence has less limitations than this BALANCER context
+    -- where the retries are executed
+    
+    -- record failure data
+    addr.failures = addr.failures or {}
+    local state, code = ngx_balancer.get_last_failure()
+    addr.failures[addr.tries-1] = { name = state, code = code }
```

Comment:
> > Regarding logging; nginx will do the actual retry check and block if it hits it.

> But since proxy_next_upstream_retries is 0 (aka no limit) by default, when exactly will "nginx block"?

If, on the first try, you call `set_more_tries(limit - 1)`, with for example limit 3. Then after the first try and maximum 2 retries, nginx will give up. See [`set_more_retries`](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#set_more_tries) 

> > So no extra logging required?

> Why are we keeping track of the failures then? Maybe don't bother? (actually need to be sure nginx will log those errors first).

This is upstream-health data, if time permits, this data can be used to disable specific dns-records after x-failures for example.